docs/en/sql-reference/functions/type-conversion-functions.md

```diff
@@ -521,6 +521,80 @@ Result:
 -   [toDate](#todate)
 -   [toDateTime](#todatetime)
 
+## parseDateTimeBestEffortUS {#parsedatetimebesteffortUS}
+
+This function is similar to ['parseDateTimeBestEffort'](#parsedatetimebesteffort), the only difference is that this function prefers US style (`MM/DD/YYYY` etc) in case of ambiguouty.
```

Comment:
> In case you decide to fix code, it will be good to fix non-US function to read US format for dates like 01/13/1970.

But for some cases, such as 12/12/1970, we don't know it's `dd/mm/yyyy` format or `mm/dd/yyyy` format, so we need ttwo functions `parseDateTimeBestEffort` and `parseDateTimeBestEffortUS` to avoid ambiguity, and user should decide which function to use.

Reply:
That's true that in such case we should have a preference because there is an ambiguity - and that's ok to have two functions for that. But when we don't have ambiguity we can either throw an error, or parse date in a single way.

To sum up all the cases:

1) `12/12/1970` - ambiguity, preference depends on function: US or non-US format.
2) `12/13/1970` - no ambiguity, either fail for non-US format or parse as US-format anyway.
3) `13/12/1970` - no ambiguity, either fail for US-format or parse as non-US-format anyway.

In #5157 it's said: "prefer US style … in case of ambiguouty", which I understand as we have to parse 2) and 3) anyway when there is no ambiguity - and not throw an error.

Also maybe @alexey-milovidov will clarify himself.

Reply:
The supported time range is from 1970 to 2105. `makeDateTime` in `base/common/DateLUTImpl.h`  makes a simple check on range of month and range of day_of_month as well as range of year, if one of them is invalid, it returns a invalid date.

Reply:
I have added some code to parse correctly when there is no ambiguity for format with separator `/, -, .`, and make a date validity check in the end. @abyss7

Reply:
That's great! Thank you.

But now you have some code style problems. Also the new build with clang-tidy shows this:

    src/IO/parseDateTimeBestEffort.cpp:547:10: error: invalid case style for local variable 'checkDate' [readability-identifier-naming,-warnings-as-errors]
    auto checkDate = [](const auto & is_leap_year_, const auto & month_, const auto & day_)
         ^~~~~~~~~
         check_date

Reply:
I have fixed the build problem and some code style problems, the last one code style problem is `BestEffortUS`, but I think `BestEffortUS` is better than `BestEffortUs` here. @abyss7

Reply:
You can just add a comment // NOLINT on the line that clang-tidy points.

Reply:
https://github.com/ClickHouse/ClickHouse/pull/12028/commits/58427e062616321c90cbe9d268cf6b2f272b96d9